### ISSUE TITLE
[Flutter] Agora-Rtc-Engine v.4.0.7 crash the app after project has been updated to support Android OS 31+

### ISSUE THREAD

**Issue by hexxyt** [AgoraIO/Agora-Flutter-SDK#504]
After we updated our **Flutter** project to support **Android OS version 31**, we updated **agora-rtc-engine** to version **4.0.7**. After this, we started experiencing crashes on Android version 31+. The reason, that they ask for Bluetooth. 
We see that Agora requires to ask Bluetooth permissions, why? We found the solution for this is to ask the Bluetooth permission (SCAN / CONNECT) each time user wants to watch/broadcast the stream, but maybe exists a more clear solution for this?

Cases crash on Android (Android OS 31+):
1. if the user has already connected Bluetooth headphones to the phone, upon watch/broadcasting the stream -> agora crash the system
2. if the user doesn't have connected Bluetooth headphones to the phone, upon watch/broadcast the stream is all ok, but if the stream is live during watching/streaming try to connect Bluetooth headphones to the phone -> agora crash the system

**Expected behavior**
Agora lib should handle permission in the library (what is used Android) and ask the user to allow permission to use Bluetooth headphones in case if they have already connected headphones (before the stream) or if a user connects the Bluetooth phones during the lives stream. 

**Smartphone (please complete the following information):**
 - OS: Android
 - Version: any of 31+

**Comment by LichKing-2234** [comment]
Any crash dump?

**Comment by hexxyt** [comment]
> Any crash dump?

**// Exception 1:**
Fatal Exception: java.lang.SecurityException: Need android.permission.BLUETOOTH_CONNECT permission for android.content.AttributionSource@d0fe668f: HeadsetService
       at android.os.Parcel.createExceptionOrNull(Parcel.java:2425)
       at android.os.Parcel.createException(Parcel.java:2409)
       at android.os.Parcel.readException(Parcel.java:2392)
       at android.os.Parcel.readException(Parcel.java:2334)
       at android.bluetooth.IBluetoothHeadset$Stub$Proxy.getConnectedDevicesWithAttribution(IBluetoothHeadset.java:1040)
       at android.bluetooth.BluetoothHeadset.getConnectedDevices(BluetoothHeadset.java:549)
       at io.agora.rtc.internal.AudioRoutingController.getBtDeviceList(AudioRoutingController.java:6)
       at io.agora.rtc.internal.AudioRoutingController.access$2100(AudioRoutingController.java)
       at io.agora.rtc.internal.AudioRoutingController$HeadsetState.headSetPlugOutProcess(AudioRoutingController.java:60)
       at io.agora.rtc.internal.AudioRoutingController$EventHandler.handleMessage(AudioRoutingController.java:166)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loopOnce(Looper.java:201)
       at android.os.Looper.loop(Looper.java:288)
       at android.app.ActivityThread.main(ActivityThread.java:7838)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)

**// Exception 2:**
Fatal Exception: java.lang.SecurityException: Need android.permission.BLUETOOTH_CONNECT permission for android.content.AttributionSource@a8ba61fa: HeadsetService
       at android.os.Parcel.createExceptionOrNull(Parcel.java:2425)
       at android.os.Parcel.createException(Parcel.java:2409)
       at android.os.Parcel.readException(Parcel.java:2392)
       at android.os.Parcel.readException(Parcel.java:2334)
       at android.bluetooth.IBluetoothHeadset$Stub$Proxy.getConnectedDevicesWithAttribution(IBluetoothHeadset.java:1040)
       at android.bluetooth.BluetoothHeadset.getConnectedDevices(BluetoothHeadset.java:549)
       at io.agora.rtc.internal.AudioRoutingController$AudioRouteBaseState.btScoDisConnectProcess(AudioRoutingController.java:15)
       at io.agora.rtc.internal.AudioRoutingController$EventHandler.handleMessage(AudioRoutingController.java:128)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loopOnce(Looper.java:201)
       at android.os.Looper.loop(Looper.java:288)
       at android.app.ActivityThread.main(ActivityThread.java:7838)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)

**Comment by LichKing-2234** [comment]
you should add `android.permission.BLUETOOTH_CONNECT` to your AndroidManifest.xml

**Comment by hexxyt** [comment]
Yes, AndroidManifest.xml has this permission. The 100% crash can be reproduced if the user watches the stream and during that tries to connect to headphones. 

So to avoid this,  need to ask ALL users to allow permission for Bluetooth connection before starting broadcasting / watching the stream, otherwise, the agora crashes the app.

How we can avoid this crash without asking all users to allow Bluetooth permission before starting using the agora rtc engine?

**Comment by hexxyt** [comment]
> you should add `android.permission.BLUETOOTH_CONNECT` to your AndroidManifest.xml

Yes, AndroidManifest.xml has this permission. The 100% crash can be reproduced if the user watches the stream and during that tries to connect to headphones.

So to avoid this, need to ask ALL users to allow permission for Bluetooth connection before starting broadcasting / watching the stream, otherwise, the agora crashes the app.

How we can avoid this crash without asking all users to allow Bluetooth permission before starting using the agora rtc engine?

**Comment by alexeylopukh-aloteq** [comment]
@LichKing-2234 I also have this issue when you specific "targetSdkVersion 31" and using Agora on Android 12 device with Bluetooth headphones. 
To fix this issue we added to the manifest: 
```
    <uses-permission android:name="android.permission.BLUETOOTH_SCAN" />
    <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
```
and before using Agora, we ask the user for these permissions (Even when no Bluetooth headphones are connected, because if he connects it while using the Agora, he will get a crash :( ).
So, this is a temporary solution. Need to fix it, because users are not happy with this solution.

**Comment by alexeylopukh-aloteq** [comment]
@LichKing-2234 
Any updates on this issue? I think it's critical.

**Comment by hexxyt** [comment]
@LichKing-2234 please let us know if you had time to take a look into this?

**Comment by LichKing-2234** [comment]
This is a known issue,  please try this solution before we fixed it.

> @LichKing-2234 I also have this issue when you specific "targetSdkVersion 31" and using Agora on Android 12 device with Bluetooth headphones. To fix this issue we added to the manifest:
> 
> ```
>     <uses-permission android:name="android.permission.BLUETOOTH_SCAN" />
>     <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
> ```
> 
> and before using Agora, we ask the user for these permissions (Even when no Bluetooth headphones are connected, because if he connects it while using the Agora, he will get a crash :( ). So, this is a temporary solution. Need to fix it, because users are not happy with this solution.

**Comment by alexeylopukh** [comment]
@LichKing-2234 Any updates on this issue?

**Comment by gabrielginter** [comment]
This is still happening, 

> to add to this, using _setEnableSpeakerphone_ has no effect if the audio is currently on the Bluetooth headphones

**Comment by littleGnAl** [comment]
This should be fixed on `agora_rtc_engine: 5.2.0`.
